segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/BaseMetadataStore.java

```diff
@@ -527,6 +542,158 @@ protected void setVersion(long version) {
+        }
+
+        private void releaseReference(String key) {
+            synchronized (keyToLockMap) {
+                LockData lockData = keyToLockMap.get(key);
+                // Decrement ref count.
+                lockData.count--;
+                // clean up if required.
+                if (0 == lockData.count) {
+                    keyToLockMap.remove(key);
+                }
+            }
+        }
+    }
+
+    /**
+     * Helper class that encapsulates lock over group of keys.
+     */
+    static class MultiKeyLock {
+        /**
+         * {@link MultiKeyLockManager} instance managing this lock.
+         */
+        private final MultiKeyLockManager lockManager;
+
+        /**
+         * Keys to lock.
+         */
+        private final String[] keysToLock;
+
+        /**
+         * Keeps track of acquired locks.
+         */
+        private final Lock[] locks;
+
+        /**
+         * Keeps track of number of locks are acquired.
+         */
+        private int aquiredLockCount = 0;
+
+        /**
+         * Whether to acquire read locks or write locks.
+         */
+        private boolean isReadOnly;
+
+        /**
+         * Constructs a {@link MultiKeyLock} instance.
+         * @param isReadOnly Whether to acquire read only locks or writable locks.
+         * @param keysToLock Keys to lock.
+         */
+        MultiKeyLock(MultiKeyLockManager lockManager, Boolean isReadOnly, String... keysToLock) {
+            Preconditions.checkState(keysToLock.length > 0, "At least one key must be locked.");
+            this.keysToLock = keysToLock;
+            this.locks = new Lock[keysToLock.length];
+            this.isReadOnly = isReadOnly;
+            this.lockManager = lockManager;
+            Arrays.sort(keysToLock);
+        }
+
+        /**
+         * Acquires all required locks.
+         */
+        void lock() {
+            for (val key: keysToLock) {
+                // Get existing lock.
+                LockData lockData = lockManager.addReference(key);
+
+                // Lock.
+                Lock lock = isReadOnly ? lockData.lock.readLock() : lockData.lock.writeLock();
+                lock.lock();
+
+                // Keep track of locks acquired.
+                locks[aquiredLockCount++] = lock;
+            }
+        }
+
+        /**
+         * Releases all acquired locks.
+         */
+        void unlock() {
+            for (int i = aquiredLockCount -1; i >= 0; i--) {
+                // Unlock.
+                locks[i].unlock();
+
+                // Update lock data.
+                lockManager.releaseReference(keysToLock[i]);
+            }
+        }
+    }
+
+    /**
+     * Data for tracking locks in {@link MultiKeyLockManager} .
+     */
+    static class LockData {
+        /**
+         * Underlying ReadWriteLock lock.
+         */
+        final ReadWriteLock lock;
+
+        /**
+         * Reference count.
+         */
+        int count;
```

Comment:
A best practice is not to expose this field to the outside (this class isn't even private). You should encapsulate whatever logic you need in this class and expose it via methods.

Reply:
fixed